### DISCUSSION TITLE
Eloquent model retrieved event triggered before relations retrieval

### DISCUSSION THREAD

**Discussion by AntoineDewaele** [laravel/framework#33003] [General]
- Laravel Version : 7.13.0
- PHP Version : 7.4

## Description

I would like to create a Trait that override `retrieved` model event to handle with a eager loaded relation.

```
trait foo
{
    public static function bootFoo()
    {
        static::retrieved(function ($model) {
            $model->relation;
        });
    }
}
```

However, it seems that `retrieved` event is fired before relation are loaded because this relation call execute another request to get the relation.

After some research, i find a discussion about it : https://github.com/laravel/framework/issues/29658 but i cannot find any follow-up or any way to get around this issue.

Is there any event that is fired after relation are eager loaded or any way to work around ?

Thanks.

**Comment by macropay-solutions** [comment]
in Eloquent Builder

    /**
     * Execute the query as a "select" statement.
     *
     * @param  array|string  $columns
     * @return \Illuminate\Database\Eloquent\Collection|static[]
     */
    public function get($columns = ['*'])
    {
        $builder = $this->applyScopes();

        // If we actually found models we will also eager load any relationships that
        // have been specified as needing to be eager loaded, which will solve the
        // n+1 query issue for the developers to avoid running a lot of queries.
        if (count($models = $builder->getModels($columns)) > 0) {
            $models = $builder->eagerLoadRelations($models);
        }

        return $builder->getModel()->newCollection($models);
    }

This function seems to be called also for the relations. So the retrieved event should be triggered for each of the related models, on the hydrate method. That is why it is not cought on the resource model.
